File: src/main/java/net/sf/rails/game/model/RightsModel.java
package net.sf.rails.game.model;

import com.google.common.base.Joiner;
import com.google.common.collect.ImmutableList;

import net.sf.rails.game.PublicCompany;
import net.sf.rails.game.special.SpecialRight;
import net.sf.rails.game.state.BooleanState;
import net.sf.rails.game.state.HashSetState;

/**
 * RightsModel stores purchased SpecialRight(s)
 *
 */
public class RightsModel extends RailsModel {

    private HashSetState<SpecialRight> rights = HashSetState.create(this, "rightsModel");

    private RightsModel(PublicCompany parent, String id) {
        super(parent, id);
    }

    public static RightsModel create(PublicCompany parent, String id) {
        return new RightsModel(parent, id);
    }

    public void add(SpecialRight right) {
        rights.add(right);
    }

    public boolean contains(SpecialRight right) {
        return rights.contains(right);
    }

    /** Get the first (normally the only) right of a certain type */
    public <T extends SpecialRight> T getRightType (Class<T> rightClass) {
        for (SpecialRight right : rights) {
            if (right.getClass() == rightClass) return (T)right;
        }
        return null;
    }

    public boolean isEmpty() { return rights.isEmpty(); }

    @Override
    public String toText() {
        StringBuilder b = new StringBuilder("");
        for (SpecialRight right:rights) {
            if (b.length() > 0) b.append(" ");
            b.append(right.toText());
        }
        return b.toString();
    }

}

File: dotnet/RailsLib.Net/Net/Game/Model/RightsModel.cs
using GameLib.Net.Game.Special;
using GameLib.Net.Game.State;
using System;
using System.Collections.Generic;
using System.Text;

/**
 * RightsModel stores purchased SpecialRight(s)
 *
 */

namespace GameLib.Net.Game.Model
{
    public class RightsModel : RailsModel
    {
        private HashSetState<SpecialRight> rights;

        private RightsModel(PublicCompany parent, string id) : base(parent, id)
        {
            rights = HashSetState<SpecialRight>.Create(this, "rightsModel");
        }

        public static RightsModel Create(PublicCompany parent, string id)
        {
            return new RightsModel(parent, id);
        }

        new public PublicCompany Parent
        {
            get
            {
                return (PublicCompany)Parent;
            }
        }

        public void Add(SpecialRight right)
        {
            rights.Add(right);
        }

        public bool Contains(SpecialRight right)
        {
            return rights.Contains(right);
        }

        override public String ToText()
        {
            List<String> rightsText = new List<string>();
            foreach (SpecialRight right in rights)
            {
                rightsText.Add(right.Name);
            }
            return string.Join(",", rightsText); //Joiner.on(",").join(rightsText.build()).toString();
        }
    }
}
